xdsl/frontend/python_code_check.py

```diff
@@ -0,0 +1,143 @@
+import ast
+from dataclasses import dataclass, field
+from typing import Dict, List, Set
+from xdsl.frontend.block import is_block
+from xdsl.frontend.exception import CodeGenerationException
+
+
+@dataclass
+class PythonCodeCheck:
+
+    @staticmethod
+    def run(stmts: List[ast.stmt]) -> None:
+        """
+        Checks if Python code within `CodeContext` is supported. On unsupported
+        cases, an exception is raised. Particularly, the check is used for
+        scoping because Python (as an interpreted language) does not have a
+        notion of constants, global variables, etc. Hence, one can redefine
+        functions, place variables in arbitrary locations and do many other
+        weird things which xDSL/MLIR would not like.
+        """
+
+        # Any code written within code context can be organized either a
+        # sequence of instructions (interpreted one by one), for example:
+        # ```
+        # with CodeContext(p):
+        #   a: i32 = 45
+        #   b: i32 = 23
+        #   result: i32 = a + b
+        # ```
+        # Alternatively, the code can be viewed as sequence of functions
+        # (possibly with a dedicated entry point), for example:
+        # ```
+        # with CodeContext(p):
+        #  def foo(x: i32) -> i32:
+        #    return x
+        #  def main():
+        #    y: i32 = foo(100)
+        # ```
+        # First, find out which of the two modes is used.
+        single_scope: bool = True
+        for stmt in stmts:
+            if isinstance(stmt, ast.FunctionDef) and not is_block(stmt):
+                single_scope = False
+                break
+
+        # Check Python code is correctly structured.
+        StructureCheck.run_with_scope(single_scope, stmts)
+
+        # TODO: Check constant/global variables are correctly defined.
+
+
+@dataclass
+class StructureCheck:
+
+    @staticmethod
+    def run_with_scope(single_scope: bool, stmts: List[ast.stmt]) -> None:
+        if single_scope:
+            visitor = SingleScopeVisitor()
+        else:
+            visitor = MultipleScopeVisitor()
+        for stmt in stmts:
+            visitor.visit(stmt)
+
+
+@dataclass
+class SingleScopeVisitor(ast.NodeVisitor):
```

Comment:
It feels like both these visitors share a lot of code/functionality. Couldn't you implement a `ScopeVisitor` and then make the `SingleScopeVisitor` and the `MultipleScopeVisitor` inherit from that?

Reply:
@webmiche The problem is that we lose nice error messages, otherwise yes, the code can be shared. I am fine with whatever.

Reply:
Sure, but only if you have the same class for both, right?

You could just change what error message is thrown in the subclass. Anyhow, if you feel like this design makes more sense, them I am fine with merging it.

Reply:
I guess we can first land this, and adapt later if necessary